File: app/src/main/java/com/example/calculator/fragments/ConverterFragment.java
package com.example.calculator.fragments;

import android.os.Bundle;
import android.os.StrictMode;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;

import androidx.fragment.app.Fragment;

import com.example.calculator.Data.ConvertData;
import com.example.calculator.Model.Measure;
import com.example.calculator.R;
import com.google.android.material.chip.ChipGroup;


public class ConverterFragment extends Fragment implements AdapterView.OnItemSelectedListener {
    View view;
    Spinner unit1, unit2;
    EditText editText1, editText2;
    TextWatcher textWatcher1, textWatcher2;
    ArrayAdapter<String> unitAdapter;
    Measure currentMeasure;
    ChipGroup chipGroup;

    public ConverterFragment() {
        // Required empty public constructor
    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        view = inflater.inflate(R.layout.fragment_converter, container, false);
        StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder().permitAll().build();

        StrictMode.setThreadPolicy(policy);
        currentMeasure = ConvertData.getLength();
        setUpSpinners();
        return view;
    }


    private void setUpSpinners() {
        unit1 = view.findViewById(R.id.unit1);
        unit2 = view.findViewById(R.id.unit2);
        editText1 = view.findViewById(R.id.editText1);
        editText2 = view.findViewById(R.id.editText2);
        chipGroup = view.findViewById(R.id.chipGroup);
        chipGroup.setOnCheckedChangeListener((group, checkedId) -> {
            if (checkedId == R.id.length) {
                currentMeasure = ConvertData.getLength();
            } else if (checkedId == R.id.mass) {
                currentMeasure = ConvertData.getMass();
            } else if (checkedId == R.id.area) {
                currentMeasure = ConvertData.getArea();
            } else if (checkedId == R.id.speed) {
                currentMeasure = ConvertData.getSpeed();
            } else if (checkedId == R.id.angle) {
                currentMeasure = ConvertData.getAngle();
            } else if (checkedId == R.id.data) {
                currentMeasure = ConvertData.getData();
            } else if (checkedId == R.id.time) {
                currentMeasure = ConvertData.getTime();
            } else if (checkedId == R.id.volume) {
                currentMeasure = ConvertData.getVolume();
            } else if (checkedId == R.id.temperature) {
                currentMeasure = ConvertData.getTemperature();
            } else if (checkedId == R.id.currency) {
                currentMeasure = ConvertData.getCurrency();
            }
            setUpSpinnerAdapter();

            textWatcher1.afterTextChanged(editText1.getText());
        });
        setUpSpinnerAdapter();


        textWatcher1 = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                double toMain, fromMain;
                if (s.toString().isEmpty()) {
                    editText2.removeTextChangedListener(textWatcher2);
                    editText2.setText("0");
                    editText2.addTextChangedListener(textWatcher2);
                    return;
                }
                double input = Double.parseDouble(s.toString());
                int selectedOne = unit1.getSelectedItemPosition();
                int selectedTwo = unit2.getSelectedItemPosition();
                if (currentMeasure.getName().equals("Temperature")) {
                    if (unit1.getSelectedItem().toString().equals("Celsius")) {
                        toMain = input;
                    } else if (unit1.getSelectedItem().toString().equals("Fahrenheit")) {
                        toMain = (input - 32) * (5 / 9.0d);
                    } else {
                        toMain = input - 273.15;
                    }
                    if (unit2.getSelectedItem().toString().equals("Celsius")) {
                        fromMain = toMain;
                    } else if (unit2.getSelectedItem().toString().equals("Fahrenheit")) {
                        fromMain = toMain * (9 / 5.0d) + 32;
                    } else {
                        fromMain = toMain + 273.15;
                    }
                } else {
                    toMain = input * currentMeasure.getToMain()[selectedOne];
                    fromMain = toMain * currentMeasure.getFromMain()[selectedTwo];
                }

                editText2.removeTextChangedListener(textWatcher2);
                String result = String.valueOf(fromMain);
                if (String.valueOf(fromMain).endsWith(".0")) {
                    editText2.setText(result.substring(0, result.length() - 2));
                } else {
                    editText2.setText(String.valueOf(fromMain));
                }
                editText2.addTextChangedListener(textWatcher2);
            }
        };
        textWatcher2 = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                double toMain, fromMain;

                if (s.toString().isEmpty()) {
                    editText1.removeTextChangedListener(textWatcher1);
                    editText1.setText("0");
                    editText1.addTextChangedListener(textWatcher1);
                    return;
                }
                double input = Double.parseDouble(s.toString());
                int selectedOne = unit1.getSelectedItemPosition();
                int selectedTwo = unit2.getSelectedItemPosition();
                if (currentMeasure.getName().equals("Temperature")) {
                    if (unit2.getSelectedItem().toString().equals("Celsius")) {
                        toMain = input;
                    } else if (unit2.getSelectedItem().toString().equals("Fahrenheit")) {
                        toMain = (input - 32) * (5 / 9.0d);
                    } else {
                        toMain = input - 273.15;
                    }
                    if (unit1.getSelectedItem().toString().equals("Celsius")) {
                        fromMain = toMain;
                    } else if (unit1.getSelectedItem().toString().equals("Fahrenheit")) {
                        fromMain = toMain * (9 / 5.0d) + 32;
                    } else {
                        fromMain = toMain + 273.15;
                    }
                } else {
                    toMain = input * currentMeasure.getToMain()[selectedTwo];
                    fromMain = toMain * currentMeasure.getFromMain()[selectedOne];
                }

                editText1.removeTextChangedListener(textWatcher1);
                String result = String.valueOf(fromMain);
                if (String.valueOf(fromMain).endsWith(".0")) {
                    editText1.setText(result.substring(0, result.length() - 2));
                } else {
                    editText1.setText(String.valueOf(fromMain));
                }
                editText1.addTextChangedListener(textWatcher1);
            }
        };

        editText1.addTextChangedListener(textWatcher1);
        editText2.addTextChangedListener(textWatcher2);


        unit1.setOnItemSelectedListener(this);
        unit2.setOnItemSelectedListener(this);


    }

    void setUpSpinnerAdapter() {
        unitAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, currentMeasure.getUnits());
        unitAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        unit1.setAdapter(unitAdapter);
        unit2.setAdapter(unitAdapter);
        unit2.setSelection(1);
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        textWatcher1.afterTextChanged(editText1.getText());
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }
}

File: app/src/main/java/com/inator/calculator/fragments/ConverterFragment.kt
package com.inator.calculator.fragments

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.inator.calculator.R
import com.inator.calculator.databinding.FragmentConverterBinding
import com.inator.calculator.viewmodel.ConverterInputViewModel

class ConverterFragment : Fragment(R.layout.fragment_converter) {
    private lateinit var textWatcher1: TextWatcher
    private lateinit var textWatcher2: TextWatcher
    private val converterInputViewModel: ConverterInputViewModel by viewModels()
    private var hasSetSavedSpinners = false

    private var _binding: FragmentConverterBinding? = null
    private val binding get() = _binding!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        _binding = FragmentConverterBinding.bind(view)
        setUpViews()
    }

    private fun setUpViews() {
        binding.chipGroup.setOnCheckedStateChangeListener { _, checkedIds ->
            when (checkedIds[0]) {
                R.id.length -> {
                    converterInputViewModel.setMeasure("Length")
                }
                R.id.mass -> {
                    converterInputViewModel.setMeasure("Mass")
                }
                R.id.area -> {
                    converterInputViewModel.setMeasure("Area")
                }
                R.id.speed -> {
                    converterInputViewModel.setMeasure("Speed")
                }
                R.id.angle -> {
                    converterInputViewModel.setMeasure("Angle")
                }
                R.id.data -> {
                    converterInputViewModel.setMeasure("Data")
                }
                R.id.time -> {
                    converterInputViewModel.setMeasure("Time")
                }
                R.id.volume -> {
                    converterInputViewModel.setMeasure("Volume")
                }
                R.id.temperature -> {
                    converterInputViewModel.setMeasure("Temperature")
                }
            }
        }
        binding.chipGroup.check(converterInputViewModel.getSavedMeasure())
        textWatcher1 = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                converterInputViewModel.setInput1(s.toString())
            }
        }
        textWatcher2 = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                converterInputViewModel.setInput2(s.toString())
            }
        }
        binding.editText1.addTextChangedListener(textWatcher1)
        binding.editText2.addTextChangedListener(textWatcher2)
        binding.unit1.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View?,
                position: Int,
                id: Long
            ) {
                converterInputViewModel.setSpinner1(position)
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}

        }
        binding.unit2.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View?,
                position: Int,
                id: Long
            ) {
                converterInputViewModel.setSpinner2(position)
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }



        converterInputViewModel.getOutputDirect().observe(viewLifecycleOwner) {
            binding.editText2.apply {
                removeTextChangedListener(textWatcher2)
                setText(it)
                addTextChangedListener(textWatcher2)
            }
        }
        converterInputViewModel.getOutputReverse().observe(viewLifecycleOwner) {
            binding.editText1.apply {
                removeTextChangedListener(textWatcher2)
                setText(it)
                addTextChangedListener(textWatcher2)
            }
        }
        converterInputViewModel.getMeasure().observe(
            viewLifecycleOwner
        ) {
            setUpSpinnerAdapter(it)
        }

    }


    private fun setUpSpinnerAdapter(string: String) {

        val unitArray = when (string) {
            "Length" -> R.array.length_units
            "Area" -> R.array.area_units
            "Mass" -> R.array.mass_units
            "Speed" -> R.array.speed_units
            "Data" -> R.array.storage_units
            "Volume" -> R.array.volume_units
            "Time" -> R.array.time_units
            "Temperature" -> R.array.temperature_units
            "Angle" -> R.array.angle_units
            else -> R.array.length_units
        }
        ArrayAdapter.createFromResource(
            requireContext(),
            unitArray,
            android.R.layout.simple_spinner_item
        ).also { unitAdapter ->
            unitAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            binding.unit1.adapter = unitAdapter
            binding.unit2.adapter = unitAdapter
            setSavedSpinners()

        }
    }

    private fun setSavedSpinners() {
        if (!hasSetSavedSpinners) {
            binding.unit1.setSelection(converterInputViewModel.getSavedSpinner1())
            binding.unit2.setSelection(converterInputViewModel.getSavedSpinner2())
            hasSetSavedSpinners = true
        } else {
            binding.unit1.setSelection(0)
            binding.unit2.setSelection(1)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}